### PR TITLE
Ma js bc offers and comps

### PR DESCRIPTION
**Pull request by feedmypixel**
First part of offers and comps done. At the moment this is setup with extra data until we get tags added to membership offers and comps.
- Pull in data from CAPI with tags ```type/quiz``` | ```tone/extraoffers```
- MembersOnlyContent Model created to extract data from the CAPI response
- Display on Offers and Comps page

#### Note we have some issues with images not being there I am not seeing assets coming from CAPI for these events although they do have images on the pages

screen shot 2015-03-12 at 15 59 09

### CODE REVIEW

File: frontend/app/model/MembersOnlyContent.scala [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import com.gu.contentapi.client.model.{Asset, Content}
+
+case class MembersOnlyContent(content: Content, assetsMain: List[Asset], fields: Option[Map[String, String]]) {
+
+  val sortedImageSizes = assetsMain.flatMap { asset =>
+    for {
+      file <- asset.file.map(_.replace("http://static", "https://static-secure"))
+      width <- asset.typeData.get("width")
+    } yield (file, width)
+  }.sortBy(_._2.toInt)
```
Comment (feedmypixel) [inline frontend/app/model/MembersOnlyContent.scala]: create List of tuples containing (file, width) then sort images by width

File: frontend/app/model/MembersOnlyContent.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  val secureThumbnail = content.fields.flatMap(_.get("secureThumbnail"))
+  val mainImg = sortedImageSizes.lastOption.fold(secureThumbnail.getOrElse(""))(_._1)
```
Comment (feedmypixel) [inline frontend/app/model/MembersOnlyContent.scala]: use large main Img if there is one if not use the secure Thumbnail image

File: frontend/app/model/MembersOnlyContent.scala [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  val srcset = sortedImageSizes.map { imageDetails =>
+    val (secureFile, width) = imageDetails
+    secureFile + " " + width + "w"
+  }.mkString(", ")
```
Comment (feedmypixel) [inline frontend/app/model/MembersOnlyContent.scala]: srcset creation - speaking with @davidrapson there may be potential of using prior art around for this

Comment (davidrapson) [inline frontend/app/model/MembersOnlyContent.scala]: Yep `ResponsiveImage.scala` abstracts this stuff. I think a lot of this code can be abstracted out; might need a little bit more refactoring to `RichEvent` to get there though.

File: frontend/app/views/offer/offersandcomps.scala.html [hunk 4]
```diff
@@ -0,0 +1,50 @@
+@(membersOnlyContent: Seq[model.MembersOnlyContent], noResultText: String)
+
+@main("Offers & Competitions") {
+    <main role="main" class="l-constrained">
+        <section class="listing-header">
+            <h1 class="listing-headline">Offers and competitions</h1>
+        </section>
+        <div class="event-blocks">
```
Comment (davidrapson) [inline frontend/app/views/offer/offersandcomps.scala.html]: Bit wary of piggybacking on the `event-` classes here. Might need to revisit this soon, either rename these classes to be more generic, or split these components out a bit more. Feels like a lot of this could just be renamed to make it more generic.

File: frontend/assets/stylesheets/components/_listings.scss [hunk 5]
```diff
@@ -2,6 +2,27 @@
...
+        padding-left: rem(gs-span(2) + ($gs-gutter * 2));
+        padding-bottom: rem($gs-gutter * 4);
+    }
+}
+
+.listing-headline {
+    @include fs-headline(4);
+
+    @include mq(tablet) {
+        @include fs-headline(5);
```
Comment (davidrapson) [inline frontend/assets/stylesheets/components/_listings.scss]: Pass `@include fs-headline(5, true)` to avoid generating duplicate font-stacks here. Not a big deal as gzip chews through this kind of repetition, but worth avoiding duplicating styles if we can.

File: frontend/assets/stylesheets/components/_listings.scss [hunk 6]
```diff
@@ -2,6 +2,27 @@
...
+    }
+    @include mq(desktop) {
+        @include fs-headline(7);
```
Comment (davidrapson) [inline frontend/assets/stylesheets/components/_listings.scss]: See above

File: frontend/app/views/offer/offersandcomps.scala.html [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+                                         srcset="@memberContent.srcset"
+                                         sizes="(min-width: 739px) 33.3vw, (min-width: 479px) 50vw, 100vw"
+                                         alt=""/>
+                                </div>
+                                <div class="event-item__content">
+                                    <div class="event-item__meta">
+                                        <h4 class="event-item__title" itemprop="name">
+                                            @fragments.inlineIcon("outbound-link", Seq("icon-inline--brand-dark"))
+                                            @if(memberContent.tagTitle) {
+                                                <span class="tone-outbound-link">@memberContent.tagTitle</span>
```
Comment (davidrapson) [inline frontend/app/views/offer/offersandcomps.scala.html]: Not sure if this should be using tone colours. Might be worth creating a new `.link-outbound` component which wraps all this up.

File: frontend/app/model/ResponsiveImage.scala [hunk 8]
```diff
@@ -7,7 +7,7 @@ case class ResponsiveImageGroup(
   altText: String,
   availableImages: Seq[ResponsiveImage]
 ) {
-  val defaultImage = availableImages.head.path
+  val defaultImage = availableImages.headOption.map(_.path)
```
Comment (davidrapson) [inline frontend/app/model/ResponsiveImage.scala]: @feedmypixel This might conflict as I've just changed this to `val defaultImage = sortedImages.find(_.width > 300).map(_.path).getOrElse(smallestImage)` in PR https://github.com/guardian/membership-frontend/pull/312

File: frontend/app/model/RichEvent.scala [hunk 9]
```diff
@@ -108,7 +108,8 @@ object RichEvent {
     }
 
     //TODO: Should be Option[String]
-    val socialImgUrl = imgOpt.map(_.defaultImage).getOrElse("")
+    //TODO in separate PR make this an option and update views
+    val socialImgUrl = imgOpt.flatMap(_.defaultImage).fold("")(d => d)
```
Comment (davidrapson) [inline frontend/app/model/RichEvent.scala]: This has also been sorted recently in `master`, it's an option now. Might want to rebase to get the latest changes.

File: frontend/assets/stylesheets/components/_article-front.scss [hunk 10]
```diff
@@ -0,0 +1,53 @@
+/* ==========================================================================
+   Items
+   ========================================================================== */
+.article-front {
```
Comment (davidrapson) [inline frontend/assets/stylesheets/components/_article-front.scss]: Could do with a comment explaining what this is.

**Comment by davidrapson** [comment]
screen shot 2015-03-12 at 14 43 26
Holy commit number batman…

**Comment by feedmypixel** [comment]
With the power of rebase = 
screen shot 2015-03-12 at 15 27 35

File: frontend/app/model/ResponsiveImage.scala (hunk 8 above)
Comment (feedmypixel) [inline frontend/app/model/ResponsiveImage.scala]: I am finding that sometimes the list of availableImages is sometimes empty hence changing this to an option

**Comment by feedmypixel** [comment]
cc @jennysivapalan can you have a glance over when you have two ticks

File: frontend/app/model/ResponsiveImage.scala (hunk 8 above)
Comment (davidrapson) [inline frontend/app/model/ResponsiveImage.scala]: @feedmypixel We shouldn't be returning a `ResponsiveImageGroup` if there are no available image sizes, so I think this is OK to no be an option.

Comment (feedmypixel) [inline frontend/app/model/ResponsiveImage.scala]: My bad - this was WIP code - reverting back agreed no need to be an option

File: frontend/app/controllers/Offer.scala [hunk 11]
```diff
@@ -2,17 +2,28 @@ package controllers
...
+import services.GuardianContentService
+
 
 trait Offer extends Controller {
   val permanentStaffGroups = Config.staffAuthorisedEmailGroups
   val AuthorisedStaff = GoogleAuthenticatedStaffAction andThen isInAuthorisedGroupGoogleAuthReq(
     permanentStaffGroups, views.html.fragments.oauth.staffWrongGroup())
+  val contentApiService = GuardianContentService
 
   // TODO move this to CachedAction once this work is ready to go into the wild
```
Comment (jennysivapalan) [inline frontend/app/controllers/Offer.scala]: Not part of this PR but might be worth removing this TODO as it's TA-DONE.

Comment (feedmypixel) [inline frontend/app/controllers/Offer.scala]: > Not part of this PR but might be worth removing this TODO as it's TA-DONE.

TWILL-DO

File: frontend/app/model/MembersOnlyContent.scala [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package model
+
+import com.gu.contentapi.client.model.{Asset, Content}
+
+case class MembersOnlyContent(content: Content) {
```
Comment (jennysivapalan) [inline frontend/app/model/MembersOnlyContent.scala]: Do you still want all this? I thought we said we'd put this in the Controller as move it if it's used in more than one place?

File: frontend/app/controllers/Offer.scala [hunk 13]
```diff
@@ -2,17 +2,28 @@ package controllers
...
   def subscriber = AuthorisedStaff { implicit request =>
     Ok(views.html.offer.subscriber())
   }
+
+  // TODO FYI this controller was originally meant for the subscribers landing page - maybe we need a rethink on the names
+  // TODO of these controllers as I feel these methods don't belong together
+  def offersAndComps = CachedAction { implicit request =>
```
Comment (jennysivapalan) [inline frontend/app/controllers/Offer.scala]: Not sure these TODOs are required?

File: frontend/app/model/MembersOnlyContent.scala (hunk 12 above)
Comment (feedmypixel) [inline frontend/app/model/MembersOnlyContent.scala]: Yeah thats right but it felt as if a lot was going on in the controller so moved it from service to model which felt a little better. But I can put it in the controller if wanted?

File: frontend/app/controllers/Offer.scala (hunk 13 above)
Comment (feedmypixel) [inline frontend/app/controllers/Offer.scala]: More of a reminder that the subscriber work and offers and Comps work doesn't feel right sat here together. 
I feel there maybe a bigger question about urls also related to this. I just created the url for subscriber so maybe @davidmcdowell needs input on that and that may filter down to the controller and where these two sit.

File: frontend/app/services/GuardianContentService.scala [hunk 14]
```diff
@@ -52,16 +52,29 @@ trait GuardianContentService extends GuardianContent {
     enumerator(Iteratee.consume()).flatMap(_.run)
   }
 
+  private def membersOnly: Future[Seq[Content]] = {
+      for {
+        response <- membersOnlyContentQuery(1)
+      } yield {
+        response.results
+      }
+  }
```
Comment (jennysivapalan) [inline frontend/app/services/GuardianContentService.scala]: You could do `membersOnlyContentQuery(1).map(_.results)` if you're getting just one page of content.

File: frontend/app/services/GuardianContentService.scala [hunk 15]
```diff
@@ -101,8 +114,23 @@ trait GuardianContent {
...
 
+  def membersOnlyContentQuery(page: Int): Future[ItemResponse] = {
+    val itemQuery = ItemQuery("extra")
+      .showFields("all")
+      .showElements("all")
+      .showTags("all")
+      .pageSize(100)
+      .page(page)
+      .tag("type/quiz | tone/extraoffers")
+    //todo: filter response for member access flag
```
Comment (jennysivapalan) [inline frontend/app/services/GuardianContentService.scala]: This filter will have to be done on the results I think rather than in the query.

File: frontend/conf/routes [hunk 16]
```diff
@@ -109,3 +109,6 @@ GET            /founder                           controllers.VanityUrl.redirect
 
 # Styleguide
 GET            /patterns                          controllers.PatternLibrary.patterns
+
+# LOL
```
Comment (jennysivapalan) [inline frontend/conf/routes]: LOL :laughing:

File: frontend/app/controllers/Offer.scala (hunk 13 above)
Comment (jennysivapalan) [inline frontend/app/controllers/Offer.scala]: Okay, we should clear up the urls before we launch.

File: frontend/conf/routes (hunk 16 above)
Comment (feedmypixel) [inline frontend/conf/routes]: @mattandrews have you been LOLing in the routes? :)

File: frontend/app/model/MembersOnlyContent.scala (hunk 12 above)
Comment (jennysivapalan) [inline frontend/app/model/MembersOnlyContent.scala]: No it's okay to leave it but the extractor below isn't necessary. I think we just need to create an apply method instead - we can pair on this if you like?

Comment (feedmypixel) [inline frontend/app/model/MembersOnlyContent.scala]: :+1: lets have a look in the morming

File: frontend/conf/routes (hunk 16 above)
Comment (mattandrews) [inline frontend/conf/routes]: Hahahaha, oops, sorry! This was just me being stupid. Forgot I committed it.

**Comment by mattandrews** [comment]
Design-wise, it sort of feels like the ones without images should have the same vertical height as the ones with images, otherwise the grid feels a bit weird.

**Comment by mattandrews** [comment]
:+1: for the code though!

**Comment by feedmypixel** [comment]
> Design-wise, it sort of feels like the ones without images should have the same vertical height as the ones with images, otherwise the grid feels a bit weird.

yeah lots of questions for this work to be answered by ben and @davidmcdowell I'd also like to have a look at this with real data to see if the lack of images is a real thing we are going to encounter or if we just need to make sure it doesn't break without an image

File: frontend/app/controllers/Offer.scala (hunk 13 above)
Comment (feedmypixel) [inline frontend/app/controllers/Offer.scala]: agreed - added to the JIRA

**Comment by feedmypixel** [comment]
@mattandrews - default image height added - needs the ok from design though
screen shot 2015-03-13 at 12 02 45

**Comment by jennysivapalan** [comment]
:+1: :o: :confetti_ball:
